benchmark/david_benchmark.py

```diff
@@ -0,0 +1,151 @@
+
+Each coreset algorithm is timed to measure and report the time taken for each step.
+"""
+
+import os
+import time
+from pathlib import Path
+from typing import Optional
+
+import cv2
+import equinox as eqx
+import jax.numpy as jnp
+import matplotlib.pyplot as plt
+import numpy as np
+from jax import random
+from mnist_benchmark import get_solver_name, initialise_solvers
+
+from coreax import Data
+
+MAX_8BIT = 255
+
+
+def downsample_opencv(image_path: str, downsampling_factor: int) -> np.ndarray:
+    """
+    Downsample an image using OpenCV resize and convert it to grayscale.
+
+    :param image_path: Path to the input image file.
+    :param downsampling_factor: Factor by which to downsample the image.
+    :return: Grayscale image after downsampling.
+    """
+    img = cv2.imread(image_path)
+
+    # Calculate new dimensions based on downsampling factor
+    scale_factor = 1 / downsampling_factor
+    width = int(img.shape[1] * scale_factor)
+    height = int(img.shape[0] * scale_factor)
+    dim = (width, height)
+    resized = cv2.resize(img, dim, interpolation=cv2.INTER_AREA)
+    # Convert to grayscale after resizing
+    grayscale_resized = cv2.cvtColor(resized, cv2.COLOR_BGR2GRAY)
+
+    return grayscale_resized
+
+
+# pylint: disable=too-many-locals
+def benchmark_coreset_algorithms(
+    in_path: Path = Path("../examples/data/david_orig.png"),
+    out_path: Optional[Path] = Path("david_benchmark_results.png"),
+    downsampling_factor: int = 2,
+):
+    """
+    Benchmark the performance of coreset algorithms on a downsampled greyscale image.
+
+    Downsample an input image, extract pixel data, and obtain coresets using various
+    algorithms. The original and coreset data are plotted and saved, and the execution
+    time is printed.
+
+    :param in_path: Path to the input image file.
+    :param out_path: Path to save the output benchmark plot image.
+    :param downsampling_factor: Factor by which to downsample the image.
+    """
+    # Base directory of the current script
+    base_dir = os.path.dirname(os.path.abspath(__file__))
+
+    # Convert to absolute paths using os.path.join
+    if not in_path.is_absolute():
+        in_path = Path(os.path.join(base_dir, in_path))
+    if out_path is not None and not out_path.is_absolute():
+        out_path = Path(os.path.join(base_dir, out_path))
+
+    original_data = downsample_opencv(str(in_path), downsampling_factor)
+    pre_coreset_data = np.column_stack(np.nonzero(original_data < MAX_8BIT))
+    pixel_values = original_data[original_data < MAX_8BIT]
+    pre_coreset_data = np.column_stack((pre_coreset_data, pixel_values)).astype(
+        np.float32
+    )
+
+    # Set up the original data object and coreset parameters
+    data = Data(pre_coreset_data)
+    coreset_size = 8_000 // (downsampling_factor**2)
+
+    # Initialize each coreset solver
+    key = random.PRNGKey(0)
+    solvers = initialise_solvers(Data(jnp.array(pre_coreset_data)), key)
```

Comment:
Why are we doing `  Data(jnp.array(pre_coreset_data))` again here? Surely this can just be:

`solvers = initialise_solvers(data, key)`. 

NB: I'm not sure which is correct for `data` - either `Data(jnp.array(pre_coreset_data))` or `Data(jnp.asarray(pre_coreset_data))`

Reply:
Have changed to `solvers = initialise_solvers(data, key)`. Both seem to work.